elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -35,8 +45,113 @@ export class SparseMerkleTree {
 	public get db(): Database {
 		return this._db;
 	}
+
+	public async getNode(nodeHash: Buffer): Promise<TreeNode> {
+		if (nodeHash.equals(EMPTY_HASH)) {
+			return new Empty();
+		}
+		const data = await this._db.get(nodeHash);
+
+		if (!data) {
+			throw new Error(
+				`Node with input hash: ${nodeHash.toString('hex')} does not exist in the tree`,
+			);
+		}
+		if (isLeaf(data)) {
+			const { key, value } = parseLeaf(data, this.keyLength);
+			return new Leaf(key, value);
+		}
+
+		const { leftHash, rightHash } = parseBranch(data);
+
+		return new Branch(leftHash, rightHash);
+	}
+	// As specified in from https://github.com/LiskHQ/lips/blob/master/proposals/lip-0039.md
+	public async update(key: Buffer, value: Buffer) {
+		if (value.length === 0) {
+			throw new Error('Value cannot be empty');
+		}
+
+		if (key.byteLength !== this.keyLength) {
+			throw new Error(`Key is not equal to defined key length of ${this.keyLength}`);
+		}
+
+		let currentNode = this.rootNode;
+		const newLeaf = new Leaf(key, value);
+		const binaryKey = binaryExpansion(key, this.keyLength);
+		// if the currentNode is EMPTY node then assign it to leafNode and return
+		if (currentNode instanceof Empty) {
+			this._rootNode = newLeaf;
+			await this._db.set(newLeaf.hash, newLeaf.data);
+			return this.rootNode;
+		}
+		const ancestorNodes: TreeNode[] = [];
+		let h = 0;
+		while (currentNode instanceof Branch) {
+			const d = binaryKey.charAt(h);
+			// Append currentNode to ancestorNodes
+			ancestorNodes.push(currentNode);
+			if (d === '0') {
+				currentNode = await this.getNode(currentNode.leftHash);
+			} else if (d === '1') {
+				currentNode = await this.getNode(currentNode.rightHash);
+			}
+			h += 1;
+		}
+
+		// The currentNode is an empty node, newLeaf will replace the default empty node or currentNode will be updated to newLeaf
+		let bottomNode: TreeNode = new Empty();
+		if (currentNode instanceof Empty) {
+			// delete the empty node and update the tree, the new leaf will substitute the empty node
+			await this._db.del(currentNode.hash);
+			bottomNode = newLeaf;
+		} else if (currentNode.key === key) {
+			bottomNode = newLeaf;
+		} else {
+			// We need to create new branches in the tree to fulfill the
+			// Condition of one leaf per empty subtree
+			// Note: h is set to the last value from the previous loop
+			const currentNodeBinaryKey = bufferToBinaryString(currentNode.key);
+			while (binaryKey.charAt(h) === currentNodeBinaryKey.charAt(h)) {
+				// Create branch node with empty value
+				const newBranch = new Branch(EMPTY_HASH, EMPTY_HASH);
+				await this._db.set(newBranch.hash, newBranch.digest);
```

Comment:
you dont have to store it in the db at this point, I would store it only once it is finalized